frame/pools/src/lib.rs

```diff
@@ -0,0 +1,1135 @@
+	points: BalanceOf<T>,
+}
+
+impl<T: Config> BondedPool<T> {
+	/// Get the amount of points to issue for some new funds that will be bonded in the pool.
+	fn points_to_issue(&self, stash: &T::AccountId, new_funds: BalanceOf<T>) -> BalanceOf<T> {
+		let bonded_balance = T::StakingInterface::bonded_balance(stash).unwrap_or(Zero::zero());
+		points_to_issue::<T>(bonded_balance, self.points, new_funds)
+	}
+
+	// Get the amount of balance to unbond from the pool based on a delegator's points of the pool.
+	fn balance_to_unbond(
+		&self,
+		stash: &T::AccountId,
+		delegator_points: BalanceOf<T>,
+	) -> BalanceOf<T> {
+		let bonded_balance = T::StakingInterface::bonded_balance(stash).unwrap_or(Zero::zero());
+		balance_to_unbond::<T>(bonded_balance, self.points, delegator_points)
+	}
+
+	// Check that the pool can accept a member with `new_funds`.
+	fn ok_to_join_with(
+		&self,
+		stash: &T::AccountId,
+		new_funds: BalanceOf<T>,
+	) -> Result<(), DispatchError> {
+		let bonded_balance = T::StakingInterface::bonded_balance(stash).unwrap_or(Zero::zero());
+		ensure!(!bonded_balance.is_zero(), Error::<T>::OverflowRisk);
+
+		let points_to_balance_ratio_floor = self
+			.points
+			// We checked for zero above
+			.div(bonded_balance);
+
+		// TODO make sure these checks make sense. Taken from staking design chat with Al
+
+		// Pool points can inflate relative to balance, but only if the pool is slashed.
+		//
+		// If we cap the ratio of points:balance so one cannot join a pool that has been slashed
+		// 90%,
+		ensure!(points_to_balance_ratio_floor < 10u32.into(), Error::<T>::OverflowRisk);
+		// while restricting the balance to 1/10th of max total issuance,
+		ensure!(
+			new_funds.saturating_add(bonded_balance) <
+				BalanceOf::<T>::max_value().div(10u32.into()),
+			Error::<T>::OverflowRisk
+		);
+		// then we can be decently confident the bonding pool points will not overflow
+		// `BalanceOf<T>`.
+		Ok(())
+	}
+}
+
+#[derive(Encode, Decode, MaxEncodedLen, TypeInfo, RuntimeDebugNoBound)]
+#[cfg_attr(feature = "std", derive(Clone, PartialEq))]
+#[codec(mel_bound(T: Config))]
+#[scale_info(skip_type_params(T))]
+pub struct RewardPool<T: Config> {
+	/// The reward destination for the pool.
+	account_id: T::AccountId,
+	/// The balance of this reward pool after the last claimed payout.
+	balance: BalanceOf<T>,
+	/// The total earnings _ever_ of this reward pool after the last claimed payout. I.E. the sum
+	/// of all incoming balance through the pools life.
+	///
+	/// NOTE: We assume this will always be less than total issuance and thus can use the runtimes
+	/// `Balance` type. However in a chain with a burn rate higher than the rate this increases,
+	/// this type should be bigger than `Balance`.
+	total_earnings: BalanceOf<T>,
+	/// The total points of this reward pool after the last claimed payout.
+	points: RewardPoints,
+}
+
+impl<T: Config> RewardPool<T> {
+	fn update_total_earnings_and_balance(mut self) -> Self {
```

Comment:
why is this not `&mut self`?

Reply:
I was trying to make it super clear whenever something is being mutated, so by consuming self here we force the user to reassign the pool with another let (or we can chain it right after querying for the pool). 
However if this is not desirable I am happy to change it

Reply:
dbb2c0bb2e9c1ed8829e20b12a5c53da4d9cc700